src/v/storage/log_manager.cc

```diff
@@ -259,20 +244,132 @@ log_manager::housekeeping_scan(model::timestamp collection_threshold) {
+
+            /*
+             * first we build a collection of ntp's as their estimated
+             * reclaimable disk space. the loop and disk_usage() are safe
+             * against concurrent log removals.
+             */
+            for (auto& log_meta : _logs_list) {
+                /*
+                 * applying segment.ms will make reclaimable data from the
+                 * active segment visible.
+                 */
+                co_await log_meta.handle.apply_segment_ms();
+                if (!log_meta.link.is_linked()) {
+                    continue;
+                }
+
+                auto ntp = log_meta.handle.config().ntp();
+                auto log = dynamic_cast<disk_log_impl*>(
+                  log_meta.handle.get_impl());
+                auto usage = co_await log->disk_usage(
+                  gc_config(collection_threshold(), _config.retention_bytes()));
+
+                /*
+                 * NOTE: this estimate is for local retention policy only. for a
+                 * policy that considers removing data uploaded to the cloud,
+                 * this estimate is available in `usage.reclaim.available`.
+                 */
+                ntp_by_gc_size.emplace(usage.reclaim.retention, std::move(ntp));
+            }
+
+            /*
+             * Since we don't hold on to a per-log gate after we've recorded it
+             * in the container above, we need to lookup the ntp by name in the
+             * official log registry to avoid problems with concurrent removals
+             * since the log interface does not tolerate ops on closed logs.
+             */
+            for (const auto& candidate : ntp_by_gc_size) {
+                auto log = get(candidate.second);
+                if (!log.has_value()) {
+                    continue;
+                }
+                co_await log->gc(
+                  gc_config(collection_threshold(), _config.retention_bytes()));
+            }
+        }
+
+        /*
+         * Fall through for an iteration of the original housekeeping loop which
+         * will perform compaction. Additional scheduling heuristics will be
+         * added here, including:
+         *
+         * - Logs can be compacted in order of most space savings first, but the
+         *   estimation will be harder, most likely based on recent compaction
+         *   ratio acehived.
+         *
+         * - It may be wise to skip compaction completely in extreme low-disk
+         *   situations because the compaction process itself requires
+         *   additional disk space to stage new segments and indices.
+         *
+         * - Enhance the `disk_usage` interface to estimate when new data will
+         *   become reclaimable and cancel non-impactful housekeeping work.
+         *
+         * - Early out compaction process if a new disk space alert arives so
+         *   that we return this main scheduling loop.
```

Comment:
May be too complex, but it might be worth considering having another fiber that always considers GCing, to give heavy compaction clusters a smoother ride, rather than relying on alerting.

Reply:
i don't think that `housekeeping_scan` (which is heavy weight) in its current form will live on for much longer. it's merely here now to deal with a single fiber being in control. we'll unify these shortly so that we can, for example, bail out on a compaction to resume GC if needed.